src/coreclr/jit/optimizebools.cpp

```diff
@@ -0,0 +1,1416 @@
+//         |   |  +--*  NE c,d
+//         |   |  \--*  GE e,f
+//         |   \--*  LT a,b
+//         \--* CNS_INT 0
+//
+//      ------------ BB03, preds={BB01} succs={BB04}
+//      *  ASG x,y
+//
+//
+//      Conditions connected by && are not yet checked for. For example:
+//          If ( a > b && c == d ) { x = y; }
+//
+bool OptBoolsDsc::optOptimizeCompareChainCondBlock()
+{
+    assert(m_b1 != nullptr && m_b2 != nullptr && m_b3 == nullptr);
+    m_t3 = nullptr;
+
+    if (!(m_b1->bbNext == m_b2 && m_b1->bbJumpDest == m_b2->bbNext && m_b1->bbJumpDest->bbNext == m_b2->bbJumpDest))
+    {
+        return false;
+    }
+
+    Statement* const s1 = optOptimizeBoolsChkBlkCond();
+    if (s1 == nullptr)
+    {
+        return false;
+    }
+    Statement* s2 = m_b2->firstStmt();
+
+    assert(m_testInfo1.testTree->OperIs(GT_JTRUE));
+    GenTree* cond1 = m_testInfo1.testTree->gtGetOp1();
+    assert(m_testInfo2.testTree->OperIs(GT_JTRUE));
+    GenTree* cond2 = m_testInfo2.testTree->gtGetOp1();
+
+    // Ensure both conditions are suitable.
+    if (!cond1->OperIsCompare() || !cond2->OperIsCompare())
+    {
+        return false;
+    }
+
+    // Ensure there are no additional side effects.
+    if ((cond1->gtFlags & (GTF_SIDE_EFFECT | GTF_ORDER_SIDEEFF)) != 0 ||
+        (cond2->gtFlags & (GTF_SIDE_EFFECT | GTF_ORDER_SIDEEFF)) != 0)
+    {
+        return false;
+    }
+
+    // Integer compares only for now (until support for Arm64 fccmp instruction is added)
+    if (varTypeIsFloating(cond1->gtGetOp1()->TypeGet()) || varTypeIsFloating(cond2->gtGetOp1()->TypeGet()))
+    {
+        return false;
+    }
+
+    // Check for previously optimized compare chains.
+    bool op1IsTestCond;
+    bool op2IsTestCond;
+    bool op1IsCondChain = FindCompareChain(cond1, &op1IsTestCond);
+    bool op2IsCondChain = FindCompareChain(cond2, &op2IsTestCond);
+    // Don't support combining multiple chains. Allowing this would give minimal benefit, as
+    // costing checks would disallow most instances.
+    if (op1IsCondChain && op2IsCondChain)
+    {
+        return false;
+    }
+
+    // Specifically for Arm64, avoid cases where optimizations in lowering will produce better
+    // code than optimizing here. Specificially:
+    // * cmp(and(...), 0) will be turned into a TEST_ opcode.
+    // * Compares against zero will be optimized with cbz.
+    if (op1IsTestCond || op2IsTestCond)
+    {
+        return false;
+    }
+
+    // Combining conditions means that all conditions are always fully evaluated.
+    // Put a limit on the max size that can be combined.
+    if (!m_comp->compStressCompile(Compiler::STRESS_OPT_BOOLS_COMPARE_CHAIN_COST, 25))
+    {
+        int op1Cost    = cond1->GetCostEx();
+        int op2Cost    = cond2->GetCostEx();
+        int maxOp1Cost = op1IsCondChain ? 35 : 7;
+        int maxOp2Cost = op2IsCondChain ? 35 : 7;
+
+        // Cost to allow for chain size of three.
+        if (op1Cost > maxOp1Cost || op2Cost > maxOp2Cost)
```

Comment:
I'm worried about a limit of `35` -- does it mean we will start unconditionally executing expensive trees that were conditional before? For example, what if the second condition contains a division?

Reply:
Yes, it will. If we were forward iterating through the blocks, then this wouldn't matter so much (an expensive condition followed by a cheap condition is probably fine). But reverse iterating means that we'll be building the chain from the last condition.
Two ways I can see around this are to either calculate the full chain depth (which I wanted to avoid). Or iterate forwards through the blocks - I'll check again to see if that's workable.


While I remember - the if conversion pass no longer needs to reverse iterate through the blocks. Once this patch is finished with, I'll fix that up.

Reply:
One of the reasons for reverse iterating through the blocks was that I was restricting myself to connecting conditions with `AND`. The final condition is always reversed in the graph, so it was easy to reverse everything else onto that.

Going forward through the blocks, this is going to require using `OR` to connect the earlier conditions. I've given this a go, and I need to double check the assembly, but it looks like lowering is correctly optimising a mixture of `OR`s and `AND`s into a valid chain of `CCMP` instructions.

>I think it's unfortunate not to support && and || with similar fidelity

Assuming the above works ok, then it should now be fairly easy to add those `&&` conditions into the pass too.